blink/blink.1
.\" Copyright 2023 Justine Alexandra Roberts Tunney
.\"
.\" Permission to use, copy, modify, and/or distribute this software for
.\" any purpose with or without fee is hereby granted, provided that the
.\" above copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
.\" WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
.\" AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
.\" DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
.\" PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
.\" TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
.\" PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd March 6, 2023
.Dt BLINK 1
.Os
.Sh NAME
.Nm blink
.Nd headless blinkenlights x86-64-linux virtual machine
.Sh SYNOPSIS
.Nm
.Op Fl hvjemZs
.Op Fl L Ar logfile
.Op Fl C Ar chroot
.Ar program
.Op Ar argv1...
.Nm
.Op Fl hvjemZs
.Op Fl L Ar logfile
.Op Fl C Ar chroot
.Fl 0
.Ar program
.Op Ar argv0...
.Sh DESCRIPTION
.Nm
is a JITing virtual machine that uses x86_64 as its byte code language
and implements the Linux Kernel ABI.
.Nm
may be used to run prebuilt binaries intended for x86_64 Linux on other
POSIX platforms.
.Sh ARGUMENTS
.Pp
The
.Ar program
argument is a
.Ev PATH
searched command, which may be:
.Pp
.Bl -dash -compact
.It
An x86_64-linux ELF executable (either static or dynamic)
.It
An Actually Portable Executable (either MZqFpD or jartsr magic)
.It
A flat executable (if
.Ar program
ends with .bin, .img, or .raw)
.It
A shell script whose #!interpreter meets the above criteria
.El
.Pp
The
.Fl 0
flag allows
.Li argv[0]
to be specified on the command line. Under normal circumstances,
.Bd -literal -offset indent
blink cmd arg1
.Ed
.Pp
is equivalent to
.Bd -literal -offset indent
execve("cmd", {"cmd", "arg1"})
.Ed
.Pp
since that's how most programs are launched. However if you
need the full power of execve() process spawning, you can say
.Bd -literal -offset indent
blink -0 cmd arg0 arg1
.Ed
.Pp
which is equivalent to
.Bd -literal -offset indent
execve("cmd", {"arg0", "arg1"})
.Ed
.Pp
.Sh OPTIONS
The following options are available:
.Bl -tag -width indent
.It Fl h
Prints condensed help.
.It Fl v
Shows
.Nm
version and build configuration details.
.It Fl e
Log to standard error (fd 2) in addition to the log file. If logging to
.Em only
standard error is desired, then
.Li -eL/dev/null
may be passed as flags.
.It Fl s
Enables system call logging (repeatable).
.Pp
This will emit to the log file
the names of system calls each time a SYSCALL instruction in executed,
along with its arguments and result. System calls are logged once
they've completed, so that the result can be shown.
.Pp
System calls are
also sometimes logged upon entry too, in which case
.Li syscall(arg) -> ...
will be logged to show that the system call has not yet completed.
Whether or not this happens depends on how many times the
.Fl s
flag is supplied.
.Pp
.Bl -dash -compact
.It
Passing
.Fl s
(once) will only log the entries of system calls that are defined as
having read + write parameters (e.g. poll, select)
.It
Passing
.Fl ss
(twice) will also log the entries of cancellation points that are likely
to block (e.g. read, accept, wait, pause).
.It
Passing
.Fl sss
(thrice) will log the entries of system calls that POSIX defines as
cancellation points but are unlikely to block (e.g. write, close, open)
which we try to avoid doing in order to reduce log noise.
.It
Passing
.Fl ssss
(4x) will log the entry of every single system call, even harmless ones
that have no business being emitted to the log twice (e.g. sigaction).
This should only be useful for the Blink dev team when the rare need
arises to troubleshoot a system call that's crashing.
.El
.Pp
System call logging isn't available in
.Li MODE=rel
and
.Li MODE=tiny
builds, in which case this flag is ignored.
.It Fl m
Enables full memory virtualization. Under normal circumstances,
.Nm
aims to share the same address space as the guest program. Depending on
the program's memory requirements (i.e. if it uses MAP_FIXED) and the
constraints imposed by the host platform, this so-called linear memory
optimization may lead to an mmap() crisis that causes Blink to not work
properly. Using this option will cause
.Nm
to virtualize the x86_64 address space precicely and reliably, however
the use of this flag carries the tradeoff of causing
.Nm
to go at least 2x slower.
.It Fl j
Disables Just-In-Time (JIT) compilation. Using this option will cause
.Nm
to go ~10x or possibly even ~100x slower.
.It Fl L Ar path
Specifies the log path. The default log path is
.Ar blink.log
in the current directory at startup. This log file won't be created
until something is actually logged. If logging to a file isn't desired,
then -L /dev/null may be used. See also the
.Fl e
flag for logging to standard error.
.It Fl C Ar path
Launch
.Ar program
in a chroot'd environment. This flag is both equivalent to and overrides
the
.Ev BLINK_OVERLAYS
environment variable.
.It Fl Z
Prints internal statistics to standard error on exit. Each line will
display a monitoring metric. Most metrics will either be integer
counters or floating point running averages. Most but not all integer
counters are monotonic. In the interest of not negatively impacting
Blink's performance, statistics are computed on a best effort basis
which currently isn't guaranteed to be atomic in a multi-threaded
environment. Statistics aren't available in MODE=rel and MODE=tiny
builds, in which case this flag is ignored.
.El
.Sh ENVIRONMENT
The following environment variables are recognized:
.Bl -tag -width indent
.It Ev BLINK_LOG_FILENAME
may be specified to supply a log path to be used in cases where the
.Fl L Ar path
flag isn't specified. This value should be an absolute path. If logging
to standard error is desired, use the
.Fl e
flag.
.It Ev BLINK_OVERLAYS
specifies one or more directories to use as the root filesystem.
Similar to
.Ev PATH
this is a colon delimited list of pathnames. If relative paths are
specified, they'll be resolved to an absolute path at startup time.
Overlays only apply to IO system calls that specify an absolute path.
The empty string overlay means use the normal / root filesystem. The
default value is
.Li :o
which means if the absolute path
.Li /$f
is opened, then first check if
.Li /$f
exists, and if it doesn't, then check if
.Li o/$f
exists, in which case open that instead. Blink uses this convention to
open shared object tests. It favors the system version if it exists, but
also downloads
.Li ld-musl-x86_64.so.1
to
.Li o/lib/ld-musl-x86_64.so.1
so the dynamic linker can transparently find it on platforms like Apple,
that don't let users put files in the root folder. On the other hand,
it's possible to say
.Li BLINK_OVERLAYS=o:
so that
.Li o/...
takes precedence over
.Li /...
(noting again that empty string means root). If a single overlay is
specified that isn't empty string, then it'll effectively act as a
restricted chroot environment.
.El
.Sh QUIRKS
Here's the current list of Blink's known quirks and tradeoffs.
.Ss "Flags"
Flag dependencies may not carry across function call boundaries under
long mode. This is because when Blink's JIT is speculating whether or
not it's necessary for an arithmetic instruction to compute flags, it
considers
.Li RET
and
.Li CALL
terminal ops that break the chain. As such
64-bit code shouldn't do things we did in the DOS days, such as using
carry flag as a return value to indicate error. This should work fine
when
.Li STC
is used to set the carry flag, but if the code computes it
cleverly using instructions like
.Li SUB
then EFLAGS might not change.
.Ss "Faults"
Blink may not report the precise program counter where a fault occurred
in
.Li ucontext_t::uc_mcontext::rip
when signalling a segmentation fault.
This is currently only possible when
.Li PUSH
or
.Li POP
access bad memory.
That's because Blink's JIT tries to avoid updating
.Li Machine::ip
on ops
it considers "pure" such as those that only access registers, which for
reasons of performance is defined to include pushing and popping.
.Ss "Threads"
Blink doesn't have a working implementation of
.Li set_robust_list()
yet,
which means robust mutexes might not get unlocked if a process crashes.
.Ss "Coherency"
POSIX.1 provides almost no guarantees of coherency, synchronization, and
durability when it comes to
.Li MAP_SHARED
mappings and recommends that
msync() be explicitly used to synchronize memory with file contents. The
Linux Kernel implements shared memory so well, that this is rarely
necessary. However some platforms like OpenBSD lack write coherency.
This means if you change a shared writable memory map and then call
pread() on the associated file region, you might get stale data. Blink
isn't able to polyfill incoherent platforms to be as coherent as Linux,
therefore apps that run in Blink should assume the POSIX rules apply.
.Ss "Signal Handling"
Blink uses
.Li SIGSYS
to deliver signals internally. This signal is
precious to Blink. It's currently not possible for guest applications to
capture it from external processes.
.Ss "Memory Protection"
Blink offers guest programs a 48-bit virtual address space with a
4096-byte page size. When programs are run on (1) host systems that have
a larger page (e.g. Apple M1, Cygwin), and (2) the linear memory
optimization is enabled (i.e. you're *not* using
.Li blink -m
) then Blink may need to relax memory protections in cases where the
memory intervals defined by the guest aren't aligned to the host system
page size. Is is recommended, when calling functions like mmap() and
mprotect(), that both
.Li addr
and
.Li addr + size
be aliged to the true page size, which Blink reports to the guest in
.Li getauxval(AT_PAGESZ).
This value should be obtainable via the portable API
.Li sysconf(_SC_PAGESIZE)
assuming the C library implements it correctly. Please note that when
Blink is running in its fully virtualized mode (i.e.
.Li blink -m
) this concern does not apply. That's because Blink will allocate a full
system page for every 4096 byte page that gets mapped from a file.
.Sh EXIT STATUS
The
.Nm
command passes along the exit code of the
.Ar program
which by convention is 0 on success or >0 on failure. In the event that
.Nm
fails to launch
.Ar program
the status 127 shall be returned.
.Sh SEE ALSO
.Xr blinkenlights 1
.Sh STANDARDS
The
.Nm
command implements a superset of the
.St -p1003.1-2008
specification, intended to emulate the behaviors of the Linux Kernel.
.Sh AUTHORS
.An "Justine Alexandra Roberts Tunney" Aq jtunney@gmail.com


blink/blink.c
/*-*- mode:c;indent-tabs-mode:nil;c-basic-offset:2;tab-width:8;coding:utf-8 -*-│
│vi: set net ft=c ts=2 sts=2 sw=2 fenc=utf-8                                :vi│
╞══════════════════════════════════════════════════════════════════════════════╡
│ Copyright 2022 Justine Alexandra Roberts Tunney                              │
│                                                                              │
│ Permission to use, copy, modify, and/or distribute this software for         │
│ any purpose with or without fee is hereby granted, provided that the         │
│ above copyright notice and this permission notice appear in all copies.      │
│                                                                              │
│ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL                │
│ WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED                │
│ WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE             │
│ AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL         │
│ DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR        │
│ PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER               │
│ TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR             │
│ PERFORMANCE OF THIS SOFTWARE.                                                │
╚─────────────────────────────────────────────────────────────────────────────*/
#include <locale.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/resource.h>
#include <unistd.h>

#include "blink/assert.h"
#include "blink/builtin.h"
#include "blink/bus.h"
#include "blink/case.h"
#include "blink/debug.h"
#include "blink/dll.h"
#include "blink/endian.h"
#include "blink/flag.h"
#include "blink/jit.h"
#include "blink/loader.h"
#include "blink/log.h"
#include "blink/machine.h"
#include "blink/macros.h"
#include "blink/map.h"
#include "blink/overlays.h"
#include "blink/pml4t.h"
#include "blink/signal.h"
#include "blink/sigwinch.h"
#include "blink/stats.h"
#include "blink/syscall.h"
#include "blink/thread.h"
#include "blink/tunables.h"
#include "blink/util.h"
#include "blink/vfs.h"
#include "blink/web.h"
#include "blink/x86.h"
#include "blink/xlat.h"

#ifndef BUILD_TIMESTAMP
#define BUILD_TIMESTAMP __TIMESTAMP__
#endif
#ifndef BUILD_MODE
#define BUILD_MODE "BUILD_MODE_UNKNOWN"
#warning "-DBUILD_MODE=... should be passed to blink/blink.c"
#endif
#ifndef BUILD_TOOLCHAIN
#define BUILD_TOOLCHAIN "BUILD_TOOLCHAIN_UNKNOWN"
#warning "-DBUILD_TOOLCHAIN=... should be passed to blink/blink.c"
#endif
#ifndef BLINK_VERSION
#define BLINK_VERSION "BLINK_VERSION_UNKNOWN"
#warning "-DBLINK_VERSION=... should be passed to blink/blink.c"
#endif
#ifndef BLINK_COMMITS
#define BLINK_COMMITS "BLINK_COMMITS_UNKNOWN"
#warning "-DBLINK_COMMITS=... should be passed to blink/blink.c"
#endif
#ifndef BLINK_GITSHA
#define BLINK_GITSHA "BLINK_GITSHA_UNKNOWN"
#warning "-DBLINK_GITSHA=... should be passed to blink/blink.c"
#endif
#ifndef CONFIG_ARGUMENTS
#define CONFIG_ARGUMENTS "CONFIG_ARGUMENTS_UNKNOWN"
#warning "-DCONFIG_ARGUMENTS=... should be passed to blink/blink.c"
#endif

#define VERSION \
  "Blink Virtual Machine " BLINK_VERSION " (" BUILD_TIMESTAMP ")\n\
Copyright (c) 2023 Justine Alexandra Roberts Tunney\n\
Blink comes with absolutely NO WARRANTY of any kind.\n\
You may redistribute copies of Blink under the ISC License.\n\
For more information, see the file named LICENSE.\n\
Toolchain: " BUILD_TOOLCHAIN "\n\
Revision: #" BLINK_COMMITS " " BLINK_GITSHA "\n\
Config: ./configure MODE=" BUILD_MODE " " CONFIG_ARGUMENTS "\n"

#define OPTS "hvjemZs0L:C:"

_Alignas(1) static const char USAGE[] =
    " [-" OPTS "] PROG [ARGS...]\n"
    "Options:\n"
    "  -h                   help\n"
#ifndef DISABLE_JIT
    "  -j                   disable jit\n"
#endif
    "  -v                   show version\n"
#ifndef NDEBUG
    "  -e                   also log to stderr\n"
#endif
    "  -0                   to specify argv[0]\n"
    "  -m                   enable memory safety\n"
#if !defined(DISABLE_STRACE) && !defined(TINY)
    "  -s                   enable system call logging\n"
#endif
#ifndef NDEBUG
    "  -Z                   print internal statistics on exit\n"
    "  -L PATH              log filename (default is blink.log)\n"
#endif
#if !defined(DISABLE_OVERLAYS) || !defined(DISABLE_VFS)
    "  -C PATH              sets chroot dir or overlay spec [default \":o\"]\n"
#endif
#if !defined(DISABLE_OVERLAYS) || !defined(NDEBUG)
    "Environment:\n"
#endif
#ifndef DISABLE_OVERLAYS
    "  $BLINK_OVERLAYS      file system roots [default \":o\"]\n"
#endif
#ifndef DISABLE_VFS
    "  $BLINK_PREFIX        file system root [default \"/\"]\n"
#endif
#ifndef NDEBUG

    "  $BLINK_LOG_FILENAME  log filename (same as -L flag)\n"
#endif
    ;

extern char **environ;
static bool FLAG_nojit;
static char g_pathbuf[PATH_MAX];

static void OnSigSys(int sig) {
  // do nothing
}

static void PrintDiagnostics(struct Machine *m) {
  ERRF("additional information\n"
       "\t%s\n"
       "%s\n"
       "%s",
       GetBacktrace(m), FormatPml4t(m), GetBlinkBacktrace());
}

void TerminateSignal(struct Machine *m, int sig, int code) {
  int syssig;
  struct sigaction sa;
  unassert(!IsSignalIgnoredByDefault(sig));
  KillOtherThreads(m->system);
#ifdef HAVE_JIT
  DisableJit(&m->system->jit);  // unmapping exec pages is slow
#endif
  if (IsSignalSerious(sig)) {
    ERRF("terminating due to %s ("
         "rip=%#" PRIx64 " "
         "code=%d "
         "faultaddr=%#" PRIx64 ")",
         DescribeSignal(sig), m->ip, code, m->faultaddr);
    PrintDiagnostics(m);
  }
  if ((syssig = XlatSignal(sig)) == -1) syssig = SIGKILL;
  FreeMachine(m);
#ifdef HAVE_JIT
  ShutdownJit();
#endif
  sa.sa_flags = 0;
  sa.sa_handler = SIG_DFL;
  sigemptyset(&sa.sa_mask);
  if (syssig != SIGKILL && syssig != SIGSTOP) {
    unassert(!sigaction(syssig, &sa, 0));
  }
  unassert(!kill(getpid(), syssig));
  Abort();
}

static void OnFatalSystemSignal(int sig, siginfo_t *si, void *ptr) {
  struct Machine *m = g_machine;
#ifdef __APPLE__
  sig = FixXnuSignal(m, sig, si);
#elif defined(__powerpc__) && CAN_64BIT
  sig = FixPpcSignal(m, sig, si);
#endif
  SIG_LOGF("OnFatalSystemSignal(%s, %p)", DescribeSignal(UnXlatSignal(sig)),
           si->si_addr);
#ifndef DISABLE_JIT
  if (IsSelfModifyingCodeSegfault(m, si)) return;
#endif
  g_siginfo = *si;
  unassert(m);
  unassert(m->canhalt);
  siglongjmp(m->onhalt, kMachineFatalSystemSignal);
}

static void ProgramLimit(struct System *s, int hresource, int gresource) {
  struct rlimit rlim;
  if (!getrlimit(hresource, &rlim)) {
    XlatRlimitToLinux(s->rlim + gresource, &rlim);
  }
}

static int Exec(char *execfn, char *prog, char **argv, char **envp) {
  int i;
  sigset_t oldmask;
  struct Machine *m, *old;
  if ((old = g_machine)) KillOtherThreads(old->system);
  unassert((g_machine = m = NewMachine(NewSystem(XED_MODE_LONG), 0)));
#ifdef HAVE_JIT
  if (FLAG_nojit) DisableJit(&m->system->jit);
#endif
  m->system->exec = Exec;
  if (!old) {
    // this is the first time a program is being loaded
    LoadProgram(m, execfn, prog, argv, envp);
    SetupCod(m);
    for (i = 0; i < 10; ++i) {
      AddStdFd(&m->system->fds, i);
    }
    ProgramLimit(m->system, RLIMIT_NOFILE, RLIMIT_NOFILE_LINUX);
  } else {
#ifdef HAVE_JIT
    DisableJit(&old->system->jit);  // unmapping exec pages is slow
#endif
    unassert(!m->sysdepth);
    unassert(!m->pagelocks.i);
    unassert(!FreeVirtual(old->system, -0x800000000000, 0x1000000000000));
    for (i = 1; i <= 64; ++i) {
      if (Read64(old->system->hands[i - 1].handler) == SIG_IGN_LINUX) {
        Write64(m->system->hands[i - 1].handler, SIG_IGN_LINUX);
      }
    }
    memcpy(m->system->rlim, old->system->rlim, sizeof(old->system->rlim));
    LoadProgram(m, execfn, prog, argv, envp);
    m->system->fds.list = old->system->fds.list;
    old->system->fds.list = 0;
    // releasing the execve() lock must come after unlocking fds
    memcpy(&oldmask, &old->system->exec_sigmask, sizeof(oldmask));
    UNLOCK(&old->system->exec_lock);
    // freeing the last machine in a system will free its system too
    FreeMachine(old);
    // restore the signal mask we had before execve() was called
    unassert(!pthread_sigmask(SIG_SETMASK, &oldmask, 0));
  }
  Blink(m);
}

static void Print(int fd, const char *s) {
  (void)write(fd, s, strlen(s));
}

_Noreturn static void PrintUsage(int argc, char *argv[], int rc, int fd) {
  Print(fd, "Usage: ");
  Print(fd, argc > 0 && argv[0] ? argv[0] : "blink");
  Print(fd, USAGE);
  exit(rc);
}

_Noreturn static void PrintVersion(void) {
  Print(1, VERSION);
  exit(0);
}

static void GetOpts(int argc, char *argv[]) {
  int opt;
  FLAG_nolinear = !CanHaveLinearMemory();
#ifndef DISABLE_OVERLAYS
  FLAG_overlays = getenv("BLINK_OVERLAYS");
  if (!FLAG_overlays) FLAG_overlays = DEFAULT_OVERLAYS;
#endif
#ifndef DISABLE_VFS
  FLAG_prefix = getenv("BLINK_PREFIX");
#endif
#if LOG_ENABLED
  FLAG_logpath = getenv("BLINK_LOG_FILENAME");
#endif
#ifdef __COSMOPOLITAN__
  if (IsWindows()) {
    FLAG_nojit = true;
    FLAG_nolinear = true;
  }
#endif
  while ((opt = GetOpt(argc, argv, OPTS)) != -1) {
    switch (opt) {
      case '0':
        FLAG_zero = true;
        break;
      case 'j':
        FLAG_nojit = true;
        break;
      case 's':
        ++FLAG_strace;
        break;
      case 'm':
        FLAG_nolinear = true;
        break;
      case 'Z':
        FLAG_statistics = true;
        break;
      case 'e':
        FLAG_alsologtostderr = true;
        break;
      case 'L':
        FLAG_logpath = optarg_;
        break;
      case 'C':
#if !defined(DISABLE_OVERLAYS)
        FLAG_overlays = optarg_;
#elif !defined(DISABLE_VFS)
        FLAG_prefix = optarg_;
#else
        WriteErrorString(
            "error: overlays and vfs support were both disabled\n");
#endif
        break;
      case 'v':
        PrintVersion();
      case 'h':
        PrintUsage(argc, argv, 0, 1);
      default:
        PrintUsage(argc, argv, 48, 2);
    }
  }
#if LOG_ENABLED
  LogInit(FLAG_logpath);
#endif
}

static void HandleSigs(void) {
  struct sigaction sa;
  signal(SIGPIPE, SIG_IGN);
  sigfillset(&sa.sa_mask);
  sa.sa_flags = 0;
#ifdef HAVE_THREADS
  sa.sa_handler = OnSigSys;
  unassert(!sigaction(SIGSYS, &sa, 0));
#endif
  sa.sa_sigaction = OnSignal;
  unassert(!sigaction(SIGINT, &sa, 0));
  unassert(!sigaction(SIGQUIT, &sa, 0));
  unassert(!sigaction(SIGHUP, &sa, 0));
  unassert(!sigaction(SIGTERM, &sa, 0));
  unassert(!sigaction(SIGXCPU, &sa, 0));
  unassert(!sigaction(SIGXFSZ, &sa, 0));
#if !defined(__SANITIZE_THREAD__) && !defined(__SANITIZE_ADDRESS__)
  sa.sa_sigaction = OnFatalSystemSignal;
  sa.sa_flags = SA_SIGINFO;
  unassert(!sigaction(SIGBUS, &sa, 0));
  unassert(!sigaction(SIGILL, &sa, 0));
  unassert(!sigaction(SIGTRAP, &sa, 0));
  unassert(!sigaction(SIGSEGV, &sa, 0));
#endif
}

#if defined(__EMSCRIPTEN__)
void exit(int status) {
  // main is called multiple times - don't perform cleanup
  _exit(status);
}
#endif

int main(int argc, char *argv[]) {
  SetupWeb();
  GetStartDir();
#ifndef DISABLE_STRACE
  setlocale(LC_ALL, "");
#endif
  // Ensure utf-8 is printed correctly on windows, this method
  // has issues(http://stackoverflow.com/a/10884364/4279) but
  // should work for at least windows 7 and newer.
#if defined(_WIN32) && !defined(__CYGWIN__)
  SetConsoleOutputCP(CP_UTF8);
#endif
  g_blink_path = argc > 0 ? argv[0] : 0;
  WriteErrorInit();
  InitMap();
  GetOpts(argc, argv);
  if (optind_ == argc) {
    PrintUsage(argc, argv, 48, 2);
  }
#ifndef DISABLE_OVERLAYS
  if (SetOverlays(FLAG_overlays, true)) {
    WriteErrorString("bad blink overlays spec; see log for details\n");
    exit(1);
  }
#endif
#ifndef DISABLE_VFS
  if (VfsInit(FLAG_prefix)) {
    WriteErrorString("error: vfs initialization failed\n");
    exit(1);
  }
#endif
  HandleSigs();
  InitBus();
  if (!Commandv(argv[optind_], g_pathbuf, sizeof(g_pathbuf))) {
    WriteErrorString(argv[0]);
    WriteErrorString(": command not found: ");
    WriteErrorString(argv[optind_]);
    WriteErrorString("\n");
    exit(127);
  }
  argv[optind_] = g_pathbuf;
  return Exec(g_pathbuf, g_pathbuf, argv + optind_ + FLAG_zero, environ);
}
